mice-and-cats-java/src/main/java/foop/world/World.java

```diff
@@ -210,19 +246,76 @@ public void entityUpdate(EntityUpdateMessage m) {
             entity.setName(m.name());
             entity.setPosition(m.position());
         } else if (m.id() == entities.size()) {
-            entities.add(new Entity(m.id(), m.name(), m.position()));
+            entities.add(new Entity(m.id(), m.name(), m.position(), m.isUnderground()));
         } else {
             throw new IllegalStateException("Unexpected entity update message: " + m);
         }
     }
 
     public void sendTo(HashSet<Player> players) {
-        var message = new GameWorldMessage(grid, subways);
+        var message = new GameWorldMessage(grid, subways.values().stream().toList());
         players.forEach(p -> p.send(message));
 
         for (Entity entity : entities) {
-            var entityUpdate = new EntityUpdateMessage(entity.getId(), entity.getName(), entity.getPosition());
+            var entityUpdate = new EntityUpdateMessage(entity.getId(), entity.getName(), entity.getPosition(), entity.isUnderground());
             players.forEach(p -> p.send(entityUpdate));
         }
     }
+
+    public void movePlayer(HashSet<Player> players, Player player, int direction) {
+
+        var entity = entities.stream().filter(e -> e.getName().equals(player.getName())).findFirst().get();
+        var position = switch (direction) {
+            case 1 -> new Position(entity.getPosition().x(), entity.getPosition().y() - 1);
+            case 2 -> new Position(entity.getPosition().x() + 1, entity.getPosition().y());
+            case 3 -> new Position(entity.getPosition().x(), entity.getPosition().y() + 1);
+            case 4 -> new Position(entity.getPosition().x() - 1, entity.getPosition().y());
+            default -> throw new IllegalArgumentException("Illegal Direction: " + direction);
+        };
+
+        boolean isMoveLegal = false;
+        boolean isUnderground = entity.isUnderground();
+
+        if (isWithinGrid(new Point(position.x(), position.y()))) {
+
+            //check if new position is an entry -> if yes, switch isUnderground
+            if (grid[position.y()][position.x()] < 0) { //axis are flipped in grid
+                isMoveLegal = true;
+                isUnderground = !entity.isUnderground();
+                log.info("is underground: " + isUnderground);
+                if (entity.isUnderground()) {
+                    //coming out of the subway
+                    //todo: switch views
+                } else {
+                    //entering a subway
+                    //todo: switch views
+                }
+
+            } else if (entity.isUnderground()) {
+                //can't leave current subway except through exit
+                var nextPoint = new Point(position.y(), position.x()); //flipped axis
+                var currPoint = new Point(entity.getPosition().y(), entity.getPosition().x());
+                var currSubway = this.cellToSubway.get(currPoint);
```

Comment:
Converting between `Point` and `Position` seems unnecessary and error prone. I'd suggest to only use `Position` in the future.

It is a bit confusing, that you swap the x and y components here. I think this is because the world generation code (`placeConnectedSubways`, `noUturn`) actually use `grid[x][y]` sometimes. Consistently indexing by `grid[row][column]` would be better.